gobblin-metastore/src/main/java/gobblin/metastore/testing/TestMetastoreDatabaseFactory.java

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright (C) 2014-2016 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed
+ * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied.
+ */
+
+package gobblin.metastore.testing;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.nio.channels.FileChannel;
+import java.nio.channels.FileLock;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.StandardOpenOption;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+
+public class TestMetastoreDatabaseFactory {
+    private static final Object syncObject = new Object();
+    private static TestMetastoreDatabase testMetastoreDatabase;
+    private static Set<ITestMetastoreDatabase> instances = new HashSet<>();
+
+    private TestMetastoreDatabaseFactory() {
+    }
+
+    public static ITestMetastoreDatabase get() throws Exception {
+        synchronized (syncObject) {
+            ensureDatabaseExists();
+            PrivateTestMetadataDatabase instance = new PrivateTestMetadataDatabase(testMetastoreDatabase);
+            instances.add(instance);
+            return instance;
+        }
+    }
+
+    static void release(ITestMetastoreDatabase instance) throws IOException {
+        synchronized (syncObject) {
+            if (instances.remove(instance) && instances.size() == 0) {
+                testMetastoreDatabase.close();
+                testMetastoreDatabase = null;
+            }
+        }
+    }
+
+    private static void ensureDatabaseExists() throws Exception {
+        if (testMetastoreDatabase == null) {
+            try (Mutex ignored = new Mutex()) {
+                if (testMetastoreDatabase == null) {
+                    testMetastoreDatabase = new TestMetastoreDatabase();
+                }
+            }
+        }
+    }
+
+    private static class Mutex implements Closeable {
```

Comment:
There should only be a single `TestMetastoreDatabase` per JVM instance. The problem is the location used to unpack the mysql tarball is common and was resulting in collisions during the travis test runs.  The mutex is to make sure that only one JVM is unpacking the embedded DB at a time.

Test use `TestMetastoreDatabaseFactory.get()` which returns a `PrivateTestMetadataDatabase`, an instance of `ITestMetastoreDatabase`.  Within one JVM, all the `PrivateTestMetadataDatabase`s reference the same `TestMetastoreDatabase`.